### PR TITLE
Early support of the Saigo toolchain

### PR DESCRIPTION
**Pull request by illwieckz**
Continuing the work done in:

- https://github.com/DaemonEngine/Daemon/pull/971

## How to use

Use this game branch too:

- https://github.com/Unvanquished/Unvanquished/pull/2862

Get the Saigo toolchain, see:

- https://github.com/DaemonEngine/Daemon/issues/795#issuecomment-1488442249

Symlink the `native_client/toolchain/linux_x86/saigo_newlib` folder into `external_deps/linux-amd64-default_10`.

Then do:

```sh
mkdir build/vms-saigo
cd build/vms-saigo
cmake ../.. -DUSE_NACL_SAIGO=ON -DBUILD_GAME_NACL=ON -DBUILD_GAME_NACL_TARGETS=native \
	-DBUILD_GAME_NATIVE_DLL=OFF -DBUILD_GAME_NATIVE_EXE=OFF \
	-DBUILD_CLIENT=OFF -DBUILD_SERVER=OFF -DBUILD_TTY_CLIENT=OFF \
&& make -j$(nproc)
```

## Current status

- ✅️: success
- ❌️: failure
- ❔️: not tested

sgame|amd64|i686|armhf
---|---|---|--
build|✅️|✅️|✅️ 
link|✅️|✅️|✅️ 
run|✅️|❔️|❔️

cgame|amd64|i686|armhf
---|---|---|--
build|✅️|✅️|✅️
link|✅️|✅️|✅️ 
run|✅️|❔️|❔️

### CODE REVIEW

**Comment by illwieckz** [comment]
Currently the sgame builds on all architectures.

I have tested the amd64 sgame and it runs.

The cgame builds on all architectures but the final linkage fails.

To fix this we may build freetype, libpng and lua with Saigo (PNaCl provided pexe `.a` for them we can't use).

It is believed this is the last thing required to acheive cgame building.

**Comment by slipher** [comment]
You can find the old repository with scripts and sometimes patches for building the PNaCl libraries [here](https://chromium.googlesource.com/webports/). Also there is a nice [summary](https://web.archive.org/web/20150802043029/https://code.google.com/p/naclports/wiki/PortList).

**Comment by illwieckz** [comment]
At first I hoped this was for building the NaCl libc. 🥲️

**Comment by illwieckz** [comment]
More should be done later. The cgame still doesn't build because we would also need to build some dependencies like zlib with Saigo, but the current status is good enough to be merged and avoid piling merge conflicts with it in the future.

File: src/common/System.cpp [hunk 1]
```diff
@@ -335,8 +337,10 @@ static void CrashHandler(const void* data, size_t n)
 
 void SetupCrashHandler()
 {
+#if !defined(__saigo__)
```
Comment (slipher) [inline src/common/System.cpp:340]: Why is this removed? It's the only debugging mechanism we have for NaCl.

File: cmake/DaemonGame.cmake [hunk 2]
```diff
@@ -75,41 +76,80 @@ function(GAMEMODULE)
...
+                    ALWAYS 1
+                )
+            endif()
         endif()
     else()
         if (FORK EQUAL 2)
-            # Put the .nexe and .pexe files in the same directory as the engine
-            set(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/..)
+            if(USE_NACL_SAIGO)
+                set(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR})
```
Comment (slipher) [inline cmake/DaemonGame.cmake:148]: Why is this different? We want them to be in the same directory as the engine so that VM type 1 can be used.

**Comment by slipher** [comment]
I think we should be able to avoid zlib by adjusting the Freetype configuration. zlib is only depended on from functions concerning gzip support. Don't think that's something we need but I could be wrong

File: src/common/System.cpp (hunk 1 above)
Comment (illwieckz) [reply src/common/System.cpp:340]: Probably because I didn't make it work.

File: cmake/DaemonGame.cmake (hunk 2 above)
Comment (illwieckz) [reply cmake/DaemonGame.cmake:154]: If my memory is right (I wrote this 1 year ago), with PNaCl we can output `sgame.pexe` to the main build directory because it's the same for all platforms, then `pnacl_finalize` translate it to `sgame-amd64.nexe`, `sgame-i686.nexe`, etc.

But with Saigo we have one subproject per platform, and it compiles to nexe directly so every subproject output an `sgame.nexe` file with the exact same name and then every platform would overwrite the file if we output it to the main build directory. So we output it the `sgame.nexe` file to every subproject build directory and then `saigo_finalize` rename them to `sgame-amd64.nexe`, `sgame-i686.nexe`, etc. into the main build directory.

You can see below that `saigo_finalize` gets `${CMAKE_RUNTIME_OUTPUT_DIRECTORY}/..` (which was to `${CMAKE_BINARY_DIR}/..`) while `pnacl_finalize` gets `${CMAKE_RUNTIME_OUTPUT_DIRECTORY}` (which was already set to `${CMAKE_BINARY_DIR}/..`).

**Comment by illwieckz** [comment]
> I think we should be able to avoid zlib by adjusting the Freetype configuration. zlib is only depended on from functions concerning gzip support. Don't think that's something we need but I could be wrong

I can build with Saigo freetype without libpng and without zlib in cgame and get it run properly.

**Comment by illwieckz** [comment]
So, this looks ready to me, it is known to be able to build a full game (cgame+sgame) that runs once other PRs taking care of other topic are merged. So these two PRs centered on the toolchain itself are good.

Improvements, like investigating and possibly re-enabling `SetupCrashHandler()` can be done later in future PRs.

File: cmake/toolchain-saigo.cmake [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+set(CMAKE_FIND_ROOT_PATH_MODE_LIBRARY ONLY)
+set(CMAKE_FIND_ROOT_PATH_MODE_PACKAGE ONLY)
+set(CMAKE_FIND_ROOT_PATH_MODE_PROGRAM BOTH)
+
+set(NACL ON)
+
+set(CMAKE_C_FLAGS "")
+set(CMAKE_CXX_FLAGS "")
+
+function(saigo_finalize dir module saigo_arch arch)
```
Comment (slipher) [inline cmake/toolchain-saigo.cmake:46]: saigo_arch is unused

File: cmake/DaemonFlags.cmake [hunk 4]
```diff
@@ -238,7 +238,9 @@ else()
 	endif()
 
 	# Optimizations.
+	if (NACL AND USE_NACL_SAIGO)
```
Comment (slipher) [inline cmake/DaemonFlags.cmake:241]: What's this change?

File: libs/zlib/gzguts.h [hunk 5]
```diff
@@ -19,6 +19,7 @@
 #endif
 
 #include <stdio.h>
+#include <unistd.h>
```
Comment (slipher) [inline libs/zlib/gzguts.h:22]: You said we can do without zlib so this is not needed, right?

File: CMakeLists.txt [hunk 6]
```diff
@@ -634,9 +634,8 @@ if ( USE_STATIC_LIBS )
     endif()
 endif()
 
-
 # zlib
-if (BUILD_CLIENT OR BUILD_TTY_CLIENT OR BUILD_SERVER OR BUILD_DUMMY_APP OR NACL)
+if (BUILD_CLIENT OR BUILD_TTY_CLIENT OR BUILD_SERVER OR BUILD_DUMMY_APP OR (NACL AND BUILD_CGAME))
```
Comment (slipher) [inline CMakeLists.txt:640]: If we don't need zlib condition could be updated to exclude Saigo builds

File: cmake/DaemonGame.cmake [hunk 7]
```diff
@@ -77,6 +77,8 @@ function(GAMEMODULE)
             endforeach(inherited_option)
```
Comment (slipher) [inline cmake/DaemonGame.cmake:77]: The variable `vm` defined a bit above is PNaCl-specific (and redundant with the newly added NACL_VMS_PROJECT).

File: cmake/DaemonFlags.cmake (hunk 4 above)
Comment (illwieckz) [reply cmake/DaemonFlags.cmake]: Something I committed by mistake when I tested:

- https://github.com/DaemonEngine/Daemon/issues/1312

Fixed.

File: libs/zlib/gzguts.h (hunk 5 above)
Comment (illwieckz) [reply libs/zlib/gzguts.h:22]: PNaCl still needs that file, but we are lucky that PNaCl doesn't require this change, so yes we can probably skip this commit, this whole folder will be removed once we drop PNaCl.

File: cmake/DaemonGame.cmake (hunk 7 above)
Comment (illwieckz) [reply cmake/DaemonGame.cmake:78]: Removed.

File: cmake/toolchain-saigo.cmake (hunk 3 above)
Comment (illwieckz) [reply cmake/toolchain-saigo.cmake]: Removed.

File: CMakeLists.txt (hunk 6 above)
Comment (illwieckz) [reply CMakeLists.txt]: Done.

File: cmake/DaemonPlatform.cmake [hunk 8]
```diff
@@ -40,4 +40,17 @@ else()
   message( FATAL_ERROR "Platform not supported" )
 endif()
 
+if (USE_NACL_SAIGO)
+	# Saigo clang reports weird errors when building some cgame and sgame arm nexe with PIE.
+	# Saigo clang crashes when building amd64 cgame with PIE, sgame builds properly though.
+	if (NACL_TARGET STREQUAL "armhf" OR NACL_TARGET STREQUAL "amd64")
```
[the remark this answers is not in the document]
Comment (illwieckz) [reply cmake/DaemonPlatform.cmake]: Done.

**Comment by slipher** [comment]
LGTM
